File: 1140-stone-game-ii/1140-stone-game-ii.py
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        n = len(piles)
        suffix_nums = [0] * (n + 1)
        for i in range(n - 1 , -1 , -1):
            suffix_nums[i] = suffix_nums[i + 1] + piles[i]
        memo = [[-1] * (n + 1) for _ in range(n)]

        def Solve(i: int , m: int) -> int:
            if i + 2 * m >= n:
                return suffix_nums[i]
            if memo[i][m] != -1:
                return memo[i][m]
            max_stone = 0
            for x in range(1 , 2*m + 1):
                opponent_score = Solve(i + x , max(m , x))
                current_score = suffix_nums[i] - opponent_score
                max_stone = max(max_stone , current_score)
            memo[i][m] = max_stone
            return memo[i][m]
        return Solve(0 , 1)

File: 1140-stone-game-ii/1140-stone-game-ii.c
#define Max(a , b) ((a) > (b) ? (a) : (b))
int Solve(int i , int m , int n , int* suffix_nums , int** memo){
    if(i + 2 * m >= n){
        return suffix_nums[i];
    }

    if(memo[i][m] != -1){
        return memo[i][m];
    }

    int max_stone = 0;
    for(int x = 1; x <= 2 * m && i + x <= n; x++){
        int opponent_score = Solve(i + x , Max(m , x) , n , suffix_nums , memo);
        int current_score = suffix_nums[i] - opponent_score;
        max_stone = Max(max_stone , current_score);
    } 

    return memo[i][m] = max_stone;
}

int stoneGameII(int* piles, int pilesSize) {
    int* suffix_nums = (int*)malloc((pilesSize + 1) * sizeof(int));
    suffix_nums[pilesSize] = 0; 
    for(int i = pilesSize - 1; i >= 0; i--){
        suffix_nums[i] = suffix_nums[i + 1] + piles[i];
    }

    int** memo = (int**)malloc(pilesSize * sizeof(int*));
    for(int i = 0; i < pilesSize; i++){
        memo[i] = (int*)malloc((pilesSize + 1) * sizeof(int));
        memset(memo[i] , -1 , (pilesSize + 1) * sizeof(int));
    }
    int result = Solve(0 , 1 , pilesSize , suffix_nums , memo);
    for(int i = 0; i < pilesSize; i++){
        free(memo[i]);
    }
    free(memo);
    free(suffix_nums);
    return result;
}